components/service/glean/src/main/java/mozilla/components/service/glean/error/ErrorRecording.kt

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+@file:Suppress("MatchingDeclarationName")
+
+package mozilla.components.service.glean.error
+
+import android.support.annotation.VisibleForTesting
+import kotlinx.coroutines.launch
+import mozilla.components.service.glean.CommonMetricData
+import mozilla.components.service.glean.CounterMetricType
+import mozilla.components.service.glean.Dispatchers
+import mozilla.components.service.glean.Lifetime
+import mozilla.components.service.glean.storages.CountersStorageEngine
+import mozilla.components.support.base.log.logger.Logger
+
+// The labeled counter metrics that store the errors are defined in
+// the `metrics.yaml` for documentation purposes, but are not actually
+// used directly, since the `sendInPings` value needs to match the
+// pings of the metric that is erroring, not some constant value that
+// we could define in `metrics.yaml`.
+
+internal enum class ErrorType {
+    OutOfRange,
+    InvalidType,
+    InvalidValue,
+    InvalidLabel
+}
+
+@Suppress("TopLevelPropertyNaming")
+internal val GLEAN_ERROR_NAMES = mapOf(
+    ErrorType.OutOfRange to "out_of_range",
+    ErrorType.InvalidType to "invalid_type",
+    ErrorType.InvalidValue to "invalid_value",
+    ErrorType.InvalidLabel to "invalid_label"
+)
+
+private var synchronousErrorRecording = false
+
+/**
+ * Make error reporting synchronous, which prevents race conditions making testing
+ * easier.
+ *
+ * @param value If true, make error reporting synchronous.  The default behavior is
+ *     asynchronous.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.NONE)
+internal fun setSynchronousErrorRecording(value: Boolean) {
+    synchronousErrorRecording = value
+}
+
+/**
+ * Record an error that will be sent as a labeled counter in the `glean.error` category
+ * in pings.
+ *
+ * @param metricData The [CommonMetricData] instance associated with the error.
+ * @param errorType The [ErrorType] type of error.
+ * @param message The message to send to `logger.warn`.  This message is not sent with the
+ *     ping.  It does not need to include the metric name, as that is automatically
+ *     prepended to the message.
+ * @param logger The [Logger] instance to display the warning.
+ */
+internal fun recordError(
+    metricData: CommonMetricData,
+    errorType: ErrorType,
+    message: String,
+    logger: Logger
+) {
+    val errorName = GLEAN_ERROR_NAMES[errorType]!!
+
+    val errorMetric = CounterMetricType(
+        disabled = false,
+        category = "glean.error",
+        lifetime = Lifetime.Ping,
+        name = "$errorName/${metricData.identifier.replace('/', '.')}",
```

Comment:
Is this the reason why we can't use the generated metrics? If so, given that we'll *always* want to not have `/` in the metirc name, can't this be in the general implementation of the storage engine? Moreover, wouldn't this still leave the `/` right after `$errorName/` in the name?

Reply:
The "/" acts as the boundary between the metric name and the label.  The error metric is a labeled metric where the labels are the name of the metric that caused the error.  That metric may itself be a labeled metric, hence the need to convert the "/" and make sure there's only one "/" anywhere in the name.

For example, if you have a labeled string metric and you do:

```
Category.MyLabeledMetric["foo"].set("too_long_string...")
```

the error should be reported using the following key in the storage engine:

```
glean.error.out_of_range/category.my_labeled_metric.foo
```

One could argue that we don't care about which specific label caused the error, in which case we would instead want to lop off the "/" and everything after it from `metricData.identifier` in the line above.  I could argue that either way.

Reply:
I see. Given the outcome of our "dots meeting", is this a problem? Do we need to watch out for `/` in keys, for this case?

Reply:
We do need to watch out for `/` in keys in the ping (and the testing we do against the schema is hopefully a guard against that).  But we still use `/` in the keys in the storage engines to markate the underlying labeled metric from the label.

So this shouldn't be a problem, but we do need to do at least one of the two options here:
- replacing `/` with `.` and thus reporting errors to the specific label for labeled metrics.
- cut the `/` and everything after it: thus reporting all errors to the "parent" labeled metric.

I'm starting to think the second option may be better (why would someone care that the error happened for search count on "Google" vs. another search engine, for example), but I don't fully understand the use cases.

Reply:
> I'm starting to think the second option may be better (why would someone care that the error happened for search count on "Google" vs. another search engine, for example), but I don't fully understand the use cases.

Yeah, plus this might get tricky privacy wise: say that we use user input as a key (we shouldn't, but we can't guarantee for the consumers of the API), and the key then we might end up sending stuff we don't want to.